src/Escher_Synth.ml

```diff
@@ -113,38 +114,54 @@ let solve_impl ?ast:(ast=false) task consts =
        List.iter ~f:(close_goal v) v_closes; true
   in
 
-  let int_components = List.filter ~f:(fun c -> c.codomain = TInt) components in
-  let bool_components = List.filter ~f:(fun c -> c.codomain = TBool) components in
-
   let apply_comp f types i =
     let rec apply_cells types acc locations = match types, locations with
-      | (typ::typs, i::locs) -> VSet.iter (fun x -> apply_cells typs (x::acc) locs) begin
-          match typ with
-            | TInt -> int_array.(i)
-            | TBool -> bool_array.(i)
-            | _ -> raise (Invalid_Type_Exn ("Escher does not handle type: "
+      | (typ::typs, i::locs) -> 
+      let apply_next = VSet.iter (fun x -> apply_cells typs (x::acc) locs) in begin
+        match typ with 
+          | TInt -> apply_next int_array.(i)
+          | TBool -> apply_next bool_array.(i)
+          | TList t -> apply_next list_array.(i)
+            (*
```

Comment:
Here I need more help. In the commented part I tried to filter from the current level of AST only those nodes whose type matches the domain of the list components. 
But for simplicity's sake I ended up doing this filtering elsewhere.